### PR TITLE
refactor(deepspeed): Refine traning code

### PR DESCRIPTION
**Pull request by xingchensong**
# Brief
- split `train.py` & `executor.py` into multiple unified API calls and move those API into `train_utils.py`
- launch `torchddp` & `deepspeed` via torchrun

# TODO (in current PR)
- [x] check training is correct (both ddp & ds)
- [x] impl model.join for deepspeed (to seamlessly support `batch_type == dynamic` and avoid to use `filter_data` before training)
- [x] log grad_norm in tensorboard for debug purpose

# TODO (in next PR)
- [ ] return dict in `processor.py::padding` to avoid make new dict in `excutor.py`
- [ ] support activation checkpointing for both ddp & deepspeed
- [ ] ssh-launcher for multi-node multi-gpu training (ssh launcher is required by torchrun)

# Warning

A known issue of deepspeed: https://github.com/microsoft/DeepSpeed/issues/4298

### CODE REVIEW

**Comment by xingchensong** [comment]
Training Pipeline:

```py
    # 1. Read config
    with open(args.config, 'r') as fin:
        configs = yaml.load(fin, Loader=yaml.FullLoader)
    if len(args.override_config) > 0:
        configs = override_config(configs, args.override_config)

    # 2. Init env for ddp OR deepspeed
    world_size, local_rank, rank = init_distributed(args)

    # 3. Do some sanity checks and save config to arsg.model_dir
    configs = check_modify_and_save_config(args, configs)

    # 4. Get dataset & dataloader
    train_dataset, cv_dataset, train_data_loader, cv_data_loader = \
        init_dataset_and_dataloader(args, configs)

    # 5. Init asr model from configs
    infos, model = init_model(args, configs)

    # 6. Check model is jitable & print model archtectures
    trace_and_print_model(model, enable_trace=True)

    # 7. Tensorboard summary
    writer = init_summarywriter(args)

    # 8. Dispatch model from cpu to gpu
    model, device = wrap_cuda_model(args, model)

    # 9. Get optimizer & scheduler
    model, optimizer, scheduler = init_optimizer_and_scheduler(
        args, infos, configs, model)

    # 10. Save checkpoints
    save_model(args, model, tag="init", infos=None)

    # 11. Get executor
    executor = init_executor(infos)

    # 12. Init scaler, used for pytorch amp mixed precision training
    scaler = None
    if args.use_amp:
        scaler = torch.cuda.amp.GradScaler()

    # 13. Start training loop
    for ...
```

**Comment by xingchensong** [comment]
To better organize arguments, this PR also split and classify different args into different category.

```py
def get_args():
    parser = argparse.ArgumentParser(description='training your network')
    parser.add_argument...
    parser.add_argument...
    parser.add_argument...
```

===>

```py
def get_args():
    parser = argparse.ArgumentParser(description='training your network')
    parser = add_model_args(parser)
    parser = add_dataset_args(parser)
    parser = add_ddp_args(parser)
    parser = add_deepspeed_args(parser)
```

**Comment by xingchensong** [comment]
Test Script:

```sh
bash run.sh --stage 4 --stop_stage 4 --data_type raw --train_set dev --train_engine torch_ddp --dir exp/conformer
```

```sh
bash run.sh --stage 4 --stop_stage 4 --data_type raw --train_set dev --train_engine deepspeed --dir exp/conformer_ds
```

**Comment by Mddct** [comment]
可以把optimizer和scheduler分开

把optimzier传给scheduler
可以方便使用者hack各种optimizer 比如lion adamw等

**Review by whiteshirt0429 — DISMISSED** [review]

**Comment by xingchensong** [comment]
Impl `join` for deepspeed:

```py
for batch_idx, batch in enumerate(data_loader):
     if wenet_join(args, device, group_join):
         break
    ...  # training step
```

where `wenet_join` is defined as:

```py
test
```

File: wenet/bin/train.py [hunk 1]
```diff
@@ -15,372 +15,119 @@
...
-                        choices=['model_only', 'model+optimizer'],
-                        help='save model/optimizer states')
-    # End deepspeed related config
-
-
-    # DeepSpeed automaticly add '--deepspeed' and '--deepspeed_config' to parser
-    parser = deepspeed.add_config_arguments(parser)
+    parser.add_argument('--train_engine',
+                        default='torch_ddp',
+                        choices=['torch_ddp', 'deepspeed', 'torch_cpu'],
```
Comment (robin1001) [inline wenet/bin/train.py:44]: cpu can also be used for parallel training with ddp.

File: wenet/bin/train.py [hunk 2]
```diff
@@ -15,372 +15,119 @@
...
+    # Dispatch model from cpu to gpu
+    model, device = wrap_cuda_model(args, model)
+
+    # Get optimizer & scheduler
+    model, optimizer, scheduler = init_optimizer_and_scheduler(
+        args, infos, configs, model)
+
+    # Save checkpoints
+    save_model(args, model, tag="init",
+               infos={"save_time": datetime.datetime.now().strftime('%d/%m/%Y %H:%M:%S')})  # noqa
```
Comment (robin1001) [inline wenet/bin/train.py:101]: use less noqa as possible as you can.

File: wenet/utils/train_utils.py [hunk 3]
```diff
@@ -0,0 +1,555 @@
...
+    if train_engine != "deepspeed":
+        return False
+
+    try:
+        # NOTE(xcsong): Why we need a new group?
+        #   Because Deepspeed has its own group where all the relevant communication
+        #   operations are executed. If we add a communication operation that is not
+        #   managed by Deepspeed in this group, it's highly likely to cause
+        #   communication chaos, resulting in hard-to-troubleshoot hangs.
+        dist.monitored_barrier(group=group_join,
```
Comment (robin1001) [inline wenet/utils/train_utils.py:401]: 除了 monitored_barrier 之外还有其他方法吗？ddp.join 是什么原理？

File: wenet/bin/train.py (hunk 1 above)
Comment (xingchensong) [reply wenet/bin/train.py:44]: okay

File: wenet/utils/train_utils.py (hunk 3 above)
Comment (xingchensong) [reply wenet/utils/train_utils.py:401]: 没细看，可能比这个要复杂点。因为观察到了这样的现象:假设两张卡，1卡有10个batch，2卡有13个batch，现有方法在2卡走到11batch时2卡发出的同步请求会出现超时（此时1卡已走完for循环）然后被捕获后手动break。而ddp的join我观察到2卡会走完完整的11，12，13batch，我猜它有机制能让2卡注意到别的卡已经完成训练，然后会取消11，12，13batch的剃度同步操作，这时候相当于单机单卡，只有2卡更新参数。然后回过来看deepspeed，因为同步操作是inside deepspeed的，我没法控制它取消同步，除非改ds源码，所以才用了当前方法

File: wenet/bin/train.py (hunk 2 above)
Comment (xingchensong) [reply wenet/bin/train.py:101]: 好的，我偷懒了😄

File: wenet/utils/train_utils.py (hunk 3 above)
Comment (robin1001) [reply wenet/utils/train_utils.py:401]: get

Comment (Mddct) [reply wenet/utils/train_utils.py:401]: 这个例子需要注意下 lr2卡的lr step 和一卡step不一样了，
下一个epoch可能造成lr不一样

这里step需要强行reduce 下（shard 分batch极度不平衡）

Comment (xingchensong) [reply wenet/utils/train_utils.py:401]: wenetjoin是在forward，backward，update parameter之前进行的，相当于，2卡在11batch的时候不会走到更新lr这一步

Comment (xingchensong) [reply wenet/utils/train_utils.py:407]: 看了一下ddp.join的说明 (https://pytorch.org/docs/1.13/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel.join) 他有两种模式：

1. `throw_on_early_termination=False`, 这时候现象就是上面举例的现象，2号卡的11,12,13batch会继续走，但是实现方式不是我猜的2号卡取消同步请求（cancel collective communication），而是1号卡增加虚拟的请求（insert “shadow” collective communication）来匹配2卡的同步请求。
2. `throw_on_early_termination=True`, 这个应该作用和本方案是一致的了

image

File: wenet/bin/train.py (hunk 1 above)
Comment (xingchensong) [reply wenet/bin/train.py:44]: 但是torchrun不支持cpu （world_size = -1）

image

干脆直接放弃cpu training这个feature把

**Comment by xingchensong** [comment]
The lines of the core files:
- `wenet/bin/train.py`: 425 -> 157
- `wenet/utils/excutor.py`: 204 -> 122

File: wenet/utils/train_utils.py (hunk 3 above)
Comment (xingchensong) [reply wenet/utils/train_utils.py:405]: `throw_on_early_termination=False` (也即pytroch joint默认方式)，的确会造成多轮训练后lr不一致

73e47b67-a759-4dff-a459-7beb7c1d20d2


`throw_on_early_termination=True` (也即这次PR joint的方式)，可以保证lr一致

image

**Review by Mddct — APPROVED** [review]

**Comment by Mddct** [comment]
deepspeech 现在还有启动hang的问题吗？

**Comment by xingchensong** [comment]
> deepspeecd 现在还有启动hang的问题吗？

没有了，服务器有多个网卡，NCCL_SOCKET_IFNAME指定网卡就可以handshake了

File: examples/aishell/s0/run.sh [hunk 4]
```diff
@@ -112,14 +115,17 @@ fi
...
-  # multi-machine training.
-  INIT_FILE=$dir/ddp_init
-  rm -f ${INIT_FILE}  # remove previous INIT_FILE
-  init_method=file://$(readlink -f $INIT_FILE)
+  init_method=tcp://$HOST_NODE_ADDR  # multi-node multi-gpu, tcp is recommended
+  if [ ${num_nodes} -le 1  ]; then
+    INIT_FILE=$dir/ddp_init
+    rm -f ${INIT_FILE}  # remove previous INIT_FILE
+    init_method=file://$(readlink -f $INIT_FILE)  # single-node single-gpu, file is recommended
+  fi
```
Comment (xingchensong) [inline examples/aishell/s0/run.sh:123]: 单机使用file启动较为稳妥，tcp方式容易出现网络问题。多机用tcp较为合适，因为file方式会hang住，怀疑是因为多机不同机器分时启动run.sh这个脚本时，init_file会冲突

ae98ff47f0b33257646e8d9232136f4

Comment (Mddct) [reply examples/aishell/s0/run.sh:123]: 多机使用file方式 在某些文件系统上 会造成死锁问题

File: wenet/bin/train.py [hunk 5]
```diff
@@ -15,407 +15,141 @@
...
+
+        dist.barrier()  # NOTE(xcsong): Ensure all ranks start Train at the same time.
         executor.train(model, optimizer, scheduler, train_data_loader, device,
-                       writer, configs, scaler)
+                       writer, configs, scaler, group_join)
+
+        dist.barrier()  # NOTE(xcsong): Ensure all ranks start CV at the same time.
         total_loss, num_seen_utts = executor.cv(model, cv_data_loader, device,
                                                 configs)
         cv_loss = total_loss / num_seen_utts
```
Comment (xingchensong) [inline wenet/bin/train.py:138]: 增加同步操作，保证不同机器的不同worker都在同一时间进入train/cv函数，这个主要是为了wenet_join更鲁棒一些。多机多卡环境下，极有可能出现某些卡更快进入下一个epoch（io快，卡性能好等原因导致），wenet_join的30s限制很容易被触发。如果能限制进入下一个epoch的时间，保持一致，那么30s超时机制会很鲁棒。对于torch_ddp，这个同步操作可有可无，加上也无妨

File: wenet/bin/train.py [hunk 6]
```diff
@@ -15,407 +15,141 @@
...
+
+    # Get optimizer & scheduler
+    model, optimizer, scheduler = init_optimizer_and_scheduler(
+        args, configs, model)
+
+    # Save checkpoints
+    save_model(model, info_dict={
+        "save_time": datetime.datetime.now().strftime('%d/%m/%Y %H:%M:%S'),
+        "tag": "init", **configs
+    })
```
Comment (xingchensong) [inline wenet/bin/train.py:104]: 旧的pipeline在保存每个epoch的yaml时，只存cvloss, step, time，这里增加了 **configs，把train_xx.yaml的内容也加进去，增加更多信息

File: wenet/utils/executor.py [hunk 7]
```diff
@@ -27,178 +30,92 @@ def __init__(self):
...
-
-    def cv(self, model, data_loader, device, args):
+                    info_dict = batch_forward(model, batch_dict, scaler, info_dict)
+                    info_dict = batch_backward(model, scaler, info_dict)
+
+                info_dict = update_parameter_and_lr(
+                    model, optimizer, scheduler,
+                    scaler, info_dict
+                )
+                log_per_step(writer, info_dict)
```
Comment (xingchensong) [inline wenet/utils/executor.py:87]: 使用 `info_dict` 串联整个过程， `batch_forward` `batch_backward` `update_parameter_and_lr` 都读入infodict，返回增加了中间信息的新infodict，信息的流动通过info_dict承载，最终在`log_per_step`中被查询和使用

File: wenet/utils/init_model.py [hunk 8]
```diff
@@ -138,4 +138,14 @@ def init_model(configs):
...
+    if args.checkpoint is not None:
+        infos = load_checkpoint(model, args.checkpoint)
+    elif args.enc_init is not None:
+        logging.info('load pretrained encoders: {}'.format(args.enc_init))
+        infos = load_trained_modules(model, args)
+    else:
+        infos = {}
+    configs["init_infos"] = infos
+    print(configs)
+    return model, configs
```
Comment (xingchensong) [inline wenet/utils/init_model.py:151]: 加载ckpt在含以上应属于init_model,因此挪到这个函数中进行，不在 `train_utils.py` 中单独起一个函数负责加载ckpt了

**Comment by kobenaxie** [comment]
重构的版本有出现显存占用变多，训练变慢的情况吗？

**Comment by xingchensong** [comment]
> 重构的版本有出现显存占用变多，训练变慢的情况吗(torch_ddp)？

有对比数据吗？是不是从某个checkpoint开始恢复训练的（如果是，那么现象估计是0号卡显存明显高于其他卡，要修改checkpoint.py的load checkpoint函数，torch.load要用cpu）训练变慢是因为torchddp也用了wenetjoin，额外增加了30s超时，小数据上可能会比较明显地增加端到端时间

**Comment by kobenaxie** [comment]
> 重构的版本有出现显存占用变多，训练变慢的情况吗(torch_ddp)？

|version |显存占用|训练时长(100 batch)|
|---------|----------|----------------------|
|Base      |15~20G |35 s                          |
|Refactor|35~40G |70 s                          |
---
我这边观察到的现象是这样的，没有加载模型，也跳过了wenet_join()，你们没有出现这个问题是吗？

**Comment by xingchensong** [comment]
> > 重构的版本有出现显存占用变多，训练变慢的情况吗(torch_ddp)？
> 
> version	显存占用	训练时长(100 batch)
> Base	15~20G	35 s
> Refactor	35~40G	70 s
> 我这边观察到的现象是这样的，没有加载模型，也跳过了wenet_join()，你们没有出现这个问题是吗？

没有，我这一直用2080ti做的实验，原来的recipe，batch最高开到16，refactor之后同样的配置是可以train的

**Comment by kobenaxie** [comment]
> > > 重构的版本有出现显存占用变多，训练变慢的情况吗(torch_ddp)？
> > 
> > 
> > version	显存占用	训练时长(100 batch)
> > Base	15~20G	35 s
> > Refactor	35~40G	70 s
> > 我这边观察到的现象是这样的，没有加载模型，也跳过了wenet_join()，你们没有出现这个问题是吗？
> 
> 没有，我这一直用2080ti做的实验，原来的recipe，batch最高开到16，refactor之后同样的配置是可以train的，方便加个微信吗，沟通更快点，微信号：currycode

好的，加你了，麻烦通过一下。

**Comment by xingchensong** [comment]
update: 8*2080ti， torch 1.13.0，测试结果显示新旧代码loss曲线和训练耗时是一致的

image

**Comment by kobenaxie** [comment]
torch从1.12.0更新到1.13.0之后显存占用跟速度正常了

**Comment by xingchensong** [comment]
@whiteshirt0429  @robin1001 @Mddct, I think this PR is ready for a final review

File: examples/aishell/s0/run.sh [hunk 9]
```diff
@@ -128,50 +127,48 @@ if [ ${stage} -le 4 ] && [ ${stop_stage} -ge 4 ]; then
...
-        --ddp.init_method $init_method \
-        --ddp.dist_backend $dist_backend \
-        --num_workers ${num_workers} \
-        --prefetch ${prefetch} \
-        $cmvn_opts \
-        --pin_memory
+    echo "$0: using torch ddp"
   fi
+
+  # NOTE(xcsong): Both ddp & deepspeed can be launched by torchrun
```
Comment (robin1001) [inline examples/aishell/s0/run.sh:136]: 这段长注释后续我们可以迁移到 docs 文档中，介绍下多机多卡如何设置和启动。

**Comment by robin1001** [comment]
Great job! 后续我们精简注释，把如何做多机多卡支持写到文档引用就行。

**Review by robin1001 — APPROVED** [review]

File: examples/aishell/s0/run.sh (hunk 9 above)
Comment (xingchensong) [reply examples/aishell/s0/run.sh:136]: okay
